nav2_behavior_tree/plugins/control/pause.cpp

```diff
@@ -0,0 +1,279 @@
+
+  // Create a separate cb group with a separate executor to spin
+  cb_group_ = node_->create_callback_group(
+    rclcpp::CallbackGroupType::MutuallyExclusive, false);
+
+  executor_ =
+    std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+
+  executor_->add_callback_group(cb_group_, node_->get_node_base_interface());
+
+  std::string pause_service_name;
+  getInput("pause_service_name", pause_service_name);
+  pause_srv_ = node_->create_service<Trigger>(
+    pause_service_name,
+    std::bind(&Pause::pause_service_callback, this, _1, _2),
+    rclcpp::ServicesQoS(), cb_group_);
+
+  std::string resume_service_name;
+  getInput("resume_service_name", resume_service_name);
+  resume_srv_ = node_->create_service<Trigger>(
+    resume_service_name,
+    std::bind(&Pause::resume_service_callback, this, _1, _2),
+    rclcpp::ServicesQoS(), cb_group_);
+
+  spinner_thread_ = std::make_unique<std::thread>(
+    [&]() {
+      executor_->spin();
+    });
+  spinner_thread_->detach();
+}
+
+Pause::~Pause()
+{
+  RCLCPP_DEBUG(node_->get_logger(), "Shutting down Pause BT node");
+  executor_->cancel();
+}
+
+BT::NodeStatus Pause::tick()
+{
+  unsigned int children_count = children_nodes_.size();
+  if (children_count < 1 || children_count > 4) {
+    throw BT::LogicError(
+            "PauseNode must have at least one and at most four children "
+            "(currently has " + std::to_string(children_count) + ")");
+  }
+
+  std::lock_guard<std::mutex> lock(state_mutex_);
+  if (status() == BT::NodeStatus::IDLE) {
+    state_ = UNPAUSED;
+  }
+  if (state_ == PAUSE_REQUESTED) {
+    resetChildren();
+    state_ = ON_PAUSE;
+    RCLCPP_INFO(node_->get_logger(), "Switched to state: ON_PAUSE");
+  }
+  if (state_ == RESUME_REQUESTED) {
+    resetChildren();
+    state_ = ON_RESUME;
+    RCLCPP_INFO(node_->get_logger(), "Switched to state: ON_RESUME");
+  }
+  setStatus(BT::NodeStatus::RUNNING);
+
+  if (state_ == ON_PAUSE) {
+    if (children_count < 3) {
+      // Event not handled, go directly to PAUSED state
+      RCLCPP_INFO(node_->get_logger(), "Switched to state: PAUSED");
+      state_ = PAUSED;
+    } else {
+      // Tick ON_PAUSE child
+      const BT::NodeStatus child_status = children_nodes_[2]->executeTick();
+      switch (child_status) {
+        case BT::NodeStatus::RUNNING:
+          break;
+        case BT::NodeStatus::SUCCESS:
+          RCLCPP_INFO(node_->get_logger(), "Switched to state: PAUSED");
+          state_ = PAUSED;
+          break;
+        case BT::NodeStatus::SKIPPED:
+          // Same as SUCCESS
+          RCLCPP_INFO(node_->get_logger(), "Switched to state: PAUSED");
+          state_ = PAUSED;
+          break;
+        case BT::NodeStatus::FAILURE:
+          RCLCPP_ERROR(node_->get_logger(), "ON_PAUSE child returned FAILURE");
+          setStatus(BT::NodeStatus::FAILURE);
+          break;
+        default:
+          throw BT::LogicError("A child node must never return IDLE");
+      }
```

Comment:
Alot of this is boilerplat between conditions, minus the log strings and the child IDX. Could this be abstracted into a function that is passed in the child to check on and the string of the state (`"PAUSED"`) to use for logging?